### ISSUE TITLE
OpenSeaDragon viewer with svs images

### ISSUE THREAD

**Issue by rita-funk** [openseadragon/openseadragon#2351]
How the OpenSeaDragon viewer supports displaying **SVS** images?

**Comment by iangilman** [comment]
OpenSeadragon can't display SVS images directly; you need to either convert them or use a server that can automatically convert them. 

I don't have any experience with SVS images, but some people use the https://iipimage.sourceforge.io/ server to serve them in a format OSD can use. Looking into that, I did run across this discussion: 

https://sourceforge.net/p/iipimage/discussion/299494/thread/36163f5546/

... Which seems to imply, at least at the time, you may still need to convert it to TIFF first. 

If you're not looking for a server, but just a conversion tool, I recommend VIPS: 

https://www.libvips.org/

https://www.libvips.org/API/current/Making-image-pyramids.html

**Comment by phillipjohnson** [comment]
@rita-funk We have had success with libvips, but we currently use the [Zoomify converter|http://zoomify.com/products.htm]. In our experience it was a little bit faster (and can run on Windows, if that is necessary).

**Comment by pearcetm** [comment]
Without having to convert the file, you can try out [the GeoTIFFTileSource (possible future plugin)](https://github.com/openseadragon/openseadragon/issues/2234) that I created using the geotiff.js library. SVS files are fundamentally TIFFs, so this tile source can work with them as native files.

**Comment by iangilman** [comment]
Oh, as long as I'm mentioning IIPImage, I should tag @ruven. Any thoughts on using IIPImage with SVS beyond the discussion I linked to above?

**Comment by ruven** [comment]
No, SVS files can't be used directly by IIPImage as they don't contain a multi-resolution pyramid. You'll need to convert to standard pyramid TIFF or if the SVS file contains an image stack, you can transcode to OME-TIFF, [which works with the latest version of IIPImage](https://iipimage.sourceforge.io/2023/03/ome-tiff-whole-slide-microscopy). Otherwise, if transcoding is not an option, the GeoTIFFTileSource plugin mentioned above looks like a good solution.

**Comment by pearcetm** [comment]
> SVS files ... don't contain a multi-resolution pyramid.

@ruven The SVS format varies a bit over time, but for the most part at least this is not true - SVS is a TIFF with a multi-resolution pyramid plus some extra associated images. See the [OpenSlide](https://openslide.org/formats/aperio/) specification page, for example. This is why they can be read using geotiff.js as mentioned above.

Now, if IIPImage doesn't support them that's a different question. But the image data is a multi-resolution pyramidal TIFF.

**Comment by ruven** [comment]
> Now, if IIPImage doesn't support them that's a different question. But the image data is a multi-resolution pyramidal TIFF.

Technically, IIPImage can and does read SVS files. I just tried with an example SVS file from the [test data](https://openslide.cs.cmu.edu/download/openslide-testdata/Aperio/) provided in your link and it can open and read it without problems. It's just that the viewer needs to handle this a little differently because of the thumbnail images and the way the pyramid is structured.

If you take the example file CMU-1.svs, tiffinfo reports the following images sizes:

```
  Image Width: 46000 Image Length: 32914 Image Depth: 1
  Image Width: 1024 Image Length: 732 Image Depth: 1
  Image Width: 11500 Image Length: 8228 Image Depth: 1
  Image Width: 2875 Image Length: 2057 Image Depth: 1
  Image Width: 387 Image Length: 463 Image Depth: 1
  Image Width: 1280 Image Length: 431 Image Depth: 1
```
Essentially there are 3 pyramid levels in the first 4 directories: Full size, 1/4 size and 1/16 size with directory 2 being the thumbnail. So, the viewer just needs to know that the resolutions are 1/4 and not 1/2 size. I'll see if I can modify IIPImage to recognize these files and discard the thumbnail images ...

For SVS JPEG2000 files, however, libtiff itself will reject them, so these are not usable with IIPImage.

**Comment by phillipjohnson** [comment]
@ruven If it is helpful, the `NewSubfileType` tag is used by Aperio on each of the TIFF directories.

- 0: Magnified slide image
- 1: Slide label
- 9: Whole slide macro view

**Comment by ruven** [comment]
[NewSubfileType = 1](https://www.awaresystems.be/imaging/tiff/tifftags/newsubfiletype.html) is normally used to indicate a reduced-resolution image and pyramid sub-resolutions are indeed tagged with this in multi-resolution TIFFs created by vips or bioformats etc. So this tag can't be used to make a reliable distinction.

However, it looks like these thumbnail images are never tiled, so maybe this could be used to distinguish between them.

**Comment by pearcetm** [comment]
> However, it looks like these thumbnail images are never tiled, so maybe this could be used to distinguish between them.

@ruven It seems that in recent versions of SVS files that some of the lower-resolution pyramid layers are also not tiled, unfortunately (I noticed this in a demo file provided by @phillipjohnson). The SVS spec is not well defined or well documented, so we're left hacking together methods of detecting them... the `ImageDescription` tag is generally useful though.

**Comment by phillipjohnson** [comment]
I made a mistake above, sorry. I went back and reviewed my notes from Leica:

> On the GT450, the Image Description value for macro and label layers empty. To detect the macro and
label layers, it is recommended to use the TIFFTAG_SUBFILETYPE tag. The macro layer will always have a
value of 9, and the label layer will always have a value of 1. This applies to both images generated by the
AT2 and the GT450.

**Comment by ruven** [comment]
I've just pushed a [commit](https://github.com/ruven/iipsrv/commit/529a091a6744276a8bb53360b4d0f4345a2d1b4c) for the IIPImage server that skips any images in the sequence that are not tiled. This is, I think, useful in general as there may be other formats that include thumbnails or other data in their multi-resolution sequence. So, the server should now handle transparently _most_ svs files. This will only work with files that are readable by libtiff, so JPEG2000-encoded tiles won't work. For atypical svs files such as that mentioned above by @pearcetm, only the tiled resolution levels will be usable.

I couldn't get OpenSeaDragon to work directly as Deepzoom etc assume each resolution level has a factor of 2 scaling (I should really get round to writing a native IIP plugin for OSD!), but here's a demo using iipmooviewer: I needed to just make a 1 line change to the code to allow arbitrary scaling factors: https://merovingio.c2rmf.cnrs.fr/iipimage/iipmooviewer/svs.html

**Comment by iangilman** [comment]
@ruven Yes, I suppose it would be good if we can support non-factor of 2 scaling. The one tile source I know that does it is the LegacyTileSource, which supports arbitrary pyramids, but no tiling on any pyramid level (each level needs to be a single tile). 

And of course, IIP plugin for OSD would be lovely!

**Comment by sbalci** [comment]
I convert svs images with vips to make a pathology atlas
https://www.patolojiatlasi.com/EN/
See template https://github.com/pathologyatlas/template and instructions https://github.com/pathologyatlas/make-html-WSI
